### PR TITLE
Cleanup ProductMultiShopRepository

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.1.x
| Description?      | https://github.com/PrestaShop/PrestaShop/issues/31373
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no (all changes of them related only to product page v2)
| Deprecations?     | no
| How to test?      | automated tests should be enough, there are no new features shipped with this PR.
| Fixed ticket?     | part of https://github.com/PrestaShop/PrestaShop/issues/31373
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

File: src/Adapter/Product/CommandHandler/DeleteProductHandler.php
Comment (zuk3975) [inline src/Adapter/Product/CommandHandler/DeleteProductHandler.php]: ping @jolelievre, you introduced DeleteProductFromShops command recently, buy I think its not used, maybe you can help me sort it out how these 2 commands should behave now?

File: src/Adapter/Product/Repository/ProductMultiShopRepository.php
Comment (zuk3975) [inline src/Adapter/Product/Repository/ProductMultiShopRepository.php]: ping @jolelievre for consultation :smile: 
this used  where we only check if product exist overall (not considering shops), so either we leave it this way, or we can make it to assertAllProductsExistsInshop and provide default product shop as argument?

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePositionFeatureContext.php
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePositionFeatureContext.php:82]: :up: just a note. Don't want to fix it in current PR, but it seems we have a method in production code only for test purposes, which I think isn't right

File: src/Adapter/Product/CommandHandler/BulkDeleteProductHandler.php
Comment (zuk3975) [inline src/Adapter/Product/CommandHandler/BulkDeleteProductHandler.php:67]: ping @jolelievre similar question related to your recently introduced DeleteProductFromShops command.

**Comment by zuk3975** [comment]
Blocked until @jolelievre (or someone else) helps me understanding what to do with those DeleteProduct and DeleteProductFromShops commands. Then I'll rename the repository and can be merged

File: src/Adapter/Product/Repository/ProductMultiShopRepository.php
Comment (FabienPapet) [inline src/Adapter/Product/Repository/ProductMultiShopRepository.php]: Doctrine + ObjectModel , on a same `AbstractMultiShopObjectModelRepository`:thinking:

There's a design problem here, doctrine usages should live inside the Prestashop bundle, not inside the adapter namespace.

This repository is already too big, we need to split here.

Comment (FabienPapet) [reply src/Adapter/Product/Repository/ProductMultiShopRepository.php]: + This repository does not have any interface.

**Review by FabienPapet — CHANGES_REQUESTED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePositionFeatureContext.php
Comment (FabienPapet) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePositionFeatureContext.php:82]: I agree with you, perhaps open another PR for this 👍

File: src/Adapter/Product/Repository/ProductMultiShopRepository.php
Comment (zuk3975) [reply src/Adapter/Product/Repository/ProductMultiShopRepository.php]: This PR only focus is to at least get rid of one of this kind of repositories for Product, because now we have 2 of them. (ProductRepository and ProductMultiShopRepository).

Most important thing that the adapter classes is not used inside core/bundle, but not what is used inside the adapter ns (which considered to be removable in future). Its kind of a dependency inversion principle I guess.
Btw Doctrine query language is only used here for more readable queries, its not like its using ORM.

I agree it is quite big already, but Im not sure how could we split it, the Product domain is big, so is the repository. Maybe we can think of something, but not in same PR I think :thinking: 

It doesn't have an interface because one of main ideas of these repositories was to wrap up legacy object models, which won't fit in any interface that could work with core rules. (I think @jolelievre explained this couple times already in some other PR's comments where you mentioned it). So I cannot redesign the whole idea of this service now.

File: src/Adapter/Product/CommandHandler/BulkDeleteProductHandler.php
Comment (jolelievre) [reply src/Adapter/Product/CommandHandler/BulkDeleteProductHandler.php:67]: So to make things clear:
- `DeleteProductCommand` for now is equivalent (or should be) to delete product from all shops
- `BulkDeleteProductCommand` is the bulk version if the first command so it's also on all shops
- `DeleteProductFromShopsCommand` was initially created for the shop association modal, it allows selecting/unselecting the shops associated to the product (see this form data handler) https://github.com/PrestaShop/PrestaShop/blob/db5a339fe735df257240c8445f9f04c6c164fba7/src/Core/Form/IdentifiableObject/DataHandler/ProductShopsFormDataHandler.php#L97
- For the controllers related to a specific shop ID or a shop group ID I totally cheated because we had no correct command and I had no time implementing them correctly at the time, so I relied `DeleteProductFromShopsCommand` to correctly fetch the related shop IDs and then remove the associations

Comment (jolelievre) [reply src/Adapter/Product/CommandHandler/BulkDeleteProductHandler.php:67]: But as you noticed, this all creates some confusion about the responsibility and the behaviour of each command, so I think it would deserve a refacto of the `DeleteProductCommand` so that it correctly handles a ShopConstraint parameter (same principle should then be applied on `BulkDeleteProductCommand`)

Although this refacto may be too big for this PR's scope so either we refacto it after or we do it before, I guess both strategies are possible.

File: src/Adapter/Product/CommandHandler/DeleteProductHandler.php
Comment (jolelievre) [reply src/Adapter/Product/CommandHandler/DeleteProductHandler.php]: As I explained in a bigger comment previously `DeleteProductFromShops` is definitely used, it was introduced to handle the shop association (removal of association to be precise) that's why it's not based on a shop constraint but on a specific list of shop IDs.

As I mentioned in the comment though, `DeleteProductCommand` would deserve a refacto to correctly handle a ShopConstraint parameter, once it's done maybe it could be used in place of the `DeleteProductFromShops` command, or maybe we should even have a `SetProductShopsCommand` that handles both association and removal in a single handler, in any case that would be a refacto to do independently after `DeleteProductCommand` has been correctly refactored.

File: src/Adapter/Product/CommandHandler/DeleteProductHandler.php
Comment (jolelievre) [inline src/Adapter/Product/CommandHandler/DeleteProductHandler.php:62]: Ah I didn't realize this method already existed, it may mean that the refacto of using a ShopConstraint here may not be that long to do. Mostly we'll need to add some behat tests to test the command with each shop constraint and update the ProductController actions to replace the hack based on `DeleteProductFromShops`.

File: src/Adapter/Product/Repository/ProductMultiShopRepository.php
Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductMultiShopRepository.php]: I agree with @zuk3975 on this one. There are flaws in our design and architecture on this kind of repository without a doubt! But I don't think it's the moment to fix them, especially considering the complexity of the product domain.

We have coming plans to improve our repositories and interfaces and the product object model repositories will actually be a good base to consider the global complex problem like:
- should we keep some "ObjectModel repositories" ? which are really handy to manipulate object models
- if we do should they implement the new repositories or be used as inline dependency to be wrapped in other "Clean interface repositories"
- what about all the "DBAL methods" that return specific data, should they return DTOs all the time, is plain array data acceptable? if so in which conditions?

All those questions need to be addressed in our new plan but this is a refacto PR that aims at removing a "duplicated repository" that was introduced to split the management of multi shop in all the product commands. We planned to merge/remove those two repositories and have only one and that's the purpose here for the 8.1.x branch before the final release.

The improvements you are referring to @FabienPapet are meant for the 9.0 version, and we'll try to implement them on the product command, actually analysing the impact they will have and the time needed to implement them all on product commands will be a good exercise to assess the viability of the new architecture.

File: src/Adapter/Product/Repository/ProductMultiShopRepository.php
Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductMultiShopRepository.php]: I think this one has a specific role to check the product existence only, not their shop association. At least in the context of related products it's all it's supposed to do. If we need a similar method to check multishop association it should be done in a dedicated method later.

File: src/PrestaShopBundle/Resources/config/services/adapter/product.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/product.yml]: Not blocking but you can probably simplify this with autowire as well

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePositionFeatureContext.php
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/UpdatePositionFeatureContext.php:82]: I also agree it should be improved in a dedicated PR but I don't know where else we should have this method for a clean separation, I guess the problem is that we don't have a dedicated query nor some easily reusable services to list products based on their position since it's mostly used in the FO for now

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

I replied most of your questions regarding the product deletion that really deserves a refacto to get rid of any confusion. However I don't if it must be done separately or not, looking at the handler and the available methods it seems adapting the `DeleteProductCommand` with a shop constraint is not too much work, but adding the related behat tests can make the whole PR grows significantly.

So if you're motivated to handle this refacto I'll let you decide wether or not you prefer to handle it separately beforhand.

File: src/Adapter/Product/CommandHandler/DeleteProductHandler.php
Comment (zuk3975) [inline src/Adapter/Product/CommandHandler/DeleteProductHandler.php:40]: @jolelievre added this todo reminder to fix these commands in dedicated PR. It would be too much for current one.

**Comment by zuk3975** [comment]
automated tests running
https://github.com/zuk3975/testing_pr/actions/runs/4233790640

**Comment by zuk3975** [comment]
automated tests https://github.com/zuk3975/testing_pr/actions/runs/4234609879

**Comment by zuk3975** [comment]
automated tests https://github.com/zuk3975/testing_pr/actions/runs/4235391566

**Comment by zuk3975** [comment]
automated tests https://github.com/zuk3975/testing_pr/actions/runs/4241420885.
Btw I tried making ProductRepository public:false, but we use it in some places (like behat contexts) with container->get() so I couldn't find a quick way to make it work and reverted it back to public.

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]:
```suggestion
```

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: This shouldn't be needed

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml]: Can you please use the injection by constructor instead, this DI by properties is not ideal as it forces having public properties and doing the same thing via constructor is not that difficult (don't forget to call `parent::_construct()`)

File: src/PrestaShopBundle/Resources/config/services/bundle/controller.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/controller.yml:161]: This one also uses public properties, but it's a flaw that was already present and we'll fix this in v9.0 (long story short this shouldn't be a controller but a rendering component)

**Review by jolelievre — CHANGES_REQUESTED** [review]

**Review by FabienPapet — DISMISSED** [review]

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975 and @FabienPapet

**Comment by jolelievre** [comment]
Running automatic campaign here https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/4244864493

**Comment by zuk3975** [comment]
Thanks @FabienPapet and @jolelievre :+1:

**Review by FabienPapet — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Hello @zuk3975 ,

@hibatallahAouadni and I have done a quick session of tests on the whole product page.

Tested the following:
- Multistore (multiple groups and multiple stores)
- Checked the BO all tabs
- Responsive BO
- Checked all types of products
- Checked FO

No exceptions have been found.
It is QA ✅ 
Thanks!

**Pull request merged**
